### PR TITLE
support sub module blame info

### PR DESCRIPTION
**Pull request by genvyday**
support blame info for sub module's file

### CODE REVIEW

**Comment by theojalba** [comment]
I am also interested in submodule blame info when using `sonar.scm.exclusions.disabled`. Will this be included in the plugin?

**Comment by pierre-guillot-sonarsource** [comment]
Hey @dbmeneses any opinion on this? :)

**Comment by pierre-guillot-sonarsource** [comment]
Hey @lukasz-jarocki-sonarsource, what's your opinion here?

File: sonar-scm-git-plugin/src/main/java/org/sonarsource/scm/git/JGitBlameCommand.java [hunk 1]
```diff
@@ -54,23 +61,43 @@ public JGitBlameCommand(PathResolver pathResolver, AnalysisWarnings analysisWarn
...
         return;
       }
+      Path parentPath=gitBaseDir.toPath();
+      Map<String,SubmoduleStatus> submap=git.submoduleStatus().call();
+	  for(Entry<String,SubmoduleStatus> e:submap.entrySet())
+	  {
+		  Path sub=parentPath.resolve(e.getKey());
+		  Repository subRepo = JGitUtils.buildRepository(sub);
+		  Git subGit=Git.wrap(subRepo);
+		  if(cloneIsInvalid(subRepo.getWorkTree())) continue;
```
Comment (lukasz-jarocki-sonarsource) [inline sonar-scm-git-plugin/src/main/java/org/sonarsource/scm/git/JGitBlameCommand.java:80]: We would need a unit test to cover this case as well.

File: sonar-scm-git-plugin/src/main/java/org/sonarsource/scm/git/JGitBlameCommand.java [hunk 2]
```diff
@@ -54,23 +61,43 @@ public JGitBlameCommand(PathResolver pathResolver, AnalysisWarnings analysisWarn
...
+      Git git = Git.wrap(repo);
       File gitBaseDir = repo.getWorkTree();
-
       if (cloneIsInvalid(gitBaseDir)) {
         return;
       }
+      Path parentPath=gitBaseDir.toPath();
+      Map<String,SubmoduleStatus> submap=git.submoduleStatus().call();
+	  for(Entry<String,SubmoduleStatus> e:submap.entrySet())
+	  {
```
Comment (lukasz-jarocki-sonarsource) [inline sonar-scm-git-plugin/src/main/java/org/sonarsource/scm/git/JGitBlameCommand.java:76]: Can you make usage of brackets consistent with rest of the codebase?

File: sonar-scm-git-plugin/src/main/java/org/sonarsource/scm/git/JGitBlameCommand.java [hunk 3]
```diff
@@ -91,15 +118,24 @@ private boolean cloneIsInvalid(File gitBaseDir) {
     return false;
   }
 
-  private void blame(BlameOutput output, Git git, File gitBaseDir, InputFile inputFile) {
+  private void blame(BlameOutput output, Map<String,Git> subGits,Git gitBase, InputFile inputFile) {
+	File gitBaseDir=gitBase.getRepository().getWorkTree();
     String filename = pathResolver.relativePath(gitBaseDir, inputFile.file());
+    String gitf=filename;
```
Comment (lukasz-jarocki-sonarsource) [inline sonar-scm-git-plugin/src/main/java/org/sonarsource/scm/git/JGitBlameCommand.java:124]: I think we can find a better name for this variable.

File: sonar-scm-git-plugin/src/main/java/org/sonarsource/scm/git/JGitBlameCommand.java [hunk 4]
```diff
@@ -54,23 +61,43 @@ public JGitBlameCommand(PathResolver pathResolver, AnalysisWarnings analysisWarn
...
       if (cloneIsInvalid(gitBaseDir)) {
         return;
       }
+      Path parentPath=gitBaseDir.toPath();
+      Map<String,SubmoduleStatus> submap=git.submoduleStatus().call();
+	  for(Entry<String,SubmoduleStatus> e:submap.entrySet())
+	  {
+		  Path sub=parentPath.resolve(e.getKey());
+		  Repository subRepo = JGitUtils.buildRepository(sub);
+		  Git subGit=Git.wrap(subRepo);
```
Comment (lukasz-jarocki-sonarsource) [inline sonar-scm-git-plugin/src/main/java/org/sonarsource/scm/git/JGitBlameCommand.java:79]: Please format the code to have consistent spacing.

**Comment by lukasz-jarocki-sonarsource** [comment]
Not really sure how much this feature is needed but there seems to be at least some interests into this. 

I think if @genvyday corrects the code style (identation and consistent brackets) and naming of the variables then we should welcome this code in our product 👍

**Comment by pierre-guillot-sonarsource** [comment]
> Not really sure how much this feature is needed but there seems to be at least some interests into this.
> 
> I think if @genvyday corrects the code style (identation and consistent brackets) and naming of the variables then we should welcome this code in our product 👍

Just so you know, git-scm is embedded in SQ since SQ 8.5. So if we want this change, we will have to apply it to SonarQube directly

**Comment by dbmeneses** [comment]
I believe the support for submodules should be complete, to support all operations we currently do with git. 
If that's not possible, we need to:
*  add tests to make sure that no operation is broken as a consequence of the SCM git plugin declaring that it supports a repository with submodules
* log when the operations are not supported, otherwise it will be very confusing why some operations are falling back to SonarQube's internal algorithms when a git repo was detected.

**Comment by genvyday** [comment]
code style and naming of the variables corrected

**Comment by wouter-admiraal-sonarsource** [comment]
This repo is now archived and read-only. We will not accept any further contributions here.

This might still be accepted and merged on https://github.com/SonarSource/sonarqube directly, which now embeds this project.

**Pull request closed without merging**